src/opf/sdpwrm.jl

```diff
@@ -0,0 +1,300 @@
+        # Other entries are not extracted as they are dense
+        primal_solution["wr"] = value.(model[:wr])
+        primal_solution["wi"] = value.(model[:wi])
+        primal_solution["pf"] = value.(model[:pf])
+        primal_solution["qf"] = value.(model[:qf])
+        primal_solution["pt"] = value.(model[:pt])
+        primal_solution["qt"] = value.(model[:qt])
+    end
+
+    return primal_solution
+end
+
+function extract_dual(opf::OPFModel{SDPOPF})
+    model = opf.model
+    T = JuMP.value_type(typeof(model))
+
+    data = opf.data
+
+    N, E, G = data.N, data.E, data.G
+    bus_fr, bus_to = data.bus_fr, data.bus_to
+
+    dual_solution = Dict{String,Any}(
+        # bus
+        "kcl_p"      => zeros(T, N),
+        "kcl_q"      => zeros(T, N),
+        "s"          => zeros(T, N),
+        # generator
+        # N/A
+        # branch
+        "ohm_pf"     => zeros(T, E),
+        "ohm_pt"     => zeros(T, E),
+        "ohm_qf"     => zeros(T, E),
+        "ohm_qt"     => zeros(T, E),
+        "va_diff"    => zeros(T, E),
+        "sm_fr"      => zeros(T, E, 3),
+        "sm_to"      => zeros(T, E, 3),
+        "sr"         => zeros(T, E),
+        "si"         => zeros(T, E),
+        # variables lower/upper bounds
+        # bus
+        "w"          => zeros(T, N),
+        # generator
+        "pg"         => zeros(T, G),
+        "qg"         => zeros(T, G),
+        # branch
+        "wr"         => zeros(T, E),
+        "wi"         => zeros(T, E),
+        "pf"         => zeros(T, E),
+        "qf"         => zeros(T, E),
+        "pt"         => zeros(T, E),
+        "qt"         => zeros(T, E),
+    )
+
+    if has_duals(model)
+        S = dual.(model[:S])
+
+        # Bus-level constraints
+        dual_solution["kcl_p"] = dual.(model[:kcl_p])
+        dual_solution["kcl_q"] = dual.(model[:kcl_q])
+        dual_solution["s"] = [S[i, i] for i in 1:N] # upper-left diagonal of S
+
+        # Generator-level constraints
+        # N/A
+
+        # Branch-level constraints
+        dual_solution["ohm_pf"] = dual.(model[:ohm_pf])
+        dual_solution["ohm_pt"] = dual.(model[:ohm_pt])
+        dual_solution["ohm_qf"] = dual.(model[:ohm_qf])
+        dual_solution["ohm_qt"] = dual.(model[:ohm_qt])
+        dual_solution["va_diff"] = dual.(model[:va_diff_lb]) + dual.(model[:va_diff_ub])  # same as bound constraints
+        dual_solution["sm_fr"] = dual.(model[:sm_fr])
+        dual_solution["sm_to"] = dual.(model[:sm_to])
+        # By construction of the PSD matrix, sr and si are defined for buspairs instead of branches.
+        # For convenience, they are stored for branches.
+        # The same values will be repeated if there are multiple branches between a bus pair.
+        dual_solution["sr"] = [S[bus_fr[e], bus_to[e]] for e in 1:E] # upper left block
+        dual_solution["si"] = [S[bus_fr[e], bus_to[e] + N] for e in 1:E] # upper right block
```

Comment:
Same here: can you give some math in the docs?
It sounds like you export a subset of the entries, from which one can presumably recover the entire matrix... how?

Reply:
Some notes on this following in-person discussion:
* The dual matrix `S` is structurally sparse (at least when we write the dual SDP in complex variables)
* Its off-diagonal non-zero entries correspond to branches, e.g., `S[i, j]` will be non-zero only if there is at least a branch between bus `i` and `j`.
* If there are two parallel branches `e1` and `e2` between buses `i` and `j`, we will extract the entry `S[i, j]` twice